### ISSUE TITLE
Wiichuck driver doesn't work with the sysfs i2cDevice

### ISSUE THREAD

**Issue by hrishikesh195** [hybridgroup/gobot#259]
When using the Wiichuck driver with a Raspberry Pi or C.H.I.P. (through the sysfs i2cDevice implementation), the following error floods the screen -

```
2016/01/31 22:15:53 Event does not exist
2016/01/31 22:15:53 Event does not exist
...
...
```

Some findings -

There are 2 parts to this problem - a) what is the event that's triggering the specific "event does not exist" error? b) why is that event getting generated?

a) The Wiichuck driver only adds the C, Z and Joystick events, but not the Error event. In `wiichuck_driver.go` -

```go
	w.AddEvent(Z)
	w.AddEvent(C)
	w.AddEvent(Joystick)
	return w
```

So whenever any kind of error is encountered (eg in I2cRead()) the above message gets generated. The fix for that is trivial. Just add `w.AddEvent(Error)` before returning.

b) The i2cDevice code in `sysfs/i2c_device.go` uses the I2C_SMBUS block transfer mode. It appears that the Wii nunchuk doesn't support it -

```
# i2cdump -y 0 0x52 s
Error: Adapter does not have SMBus block read capability
```

The thing to note here is that this only affects I2cRead() for the wiichuck. Although I2cWrite is also called, it is only called for writing 1 or 2 bytes and for those lengths, the code doesn't do I2C_SMBUS but instead writes directly on the file descriptor -

```go
func (d *i2cDevice) Write(b []byte) (n int, err error) {
	if len(b) <= 2 {
		return d.file.Write(b)
	}
...
```

Looking at the relevant i2c header file in linux kernel source (http://lxr.free-electrons.com/source/include/uapi/linux/i2c-dev.h#L47) it appears that there is an ioctl command to get the functionality (I2C_FUNCS). So a possible way to fix this would be to first query the [functionality supported](http://lxr.free-electrons.com/source/include/uapi/linux/i2c.h#L83) by the specific adapter and then pick an appropriate transfer mode.

**Comment by hrishikesh195** [comment]
It turns out that the ioctl for I2C_FUNCS doesn't take in any argument identifying the specific device address. So the available I2C functionality depends on the hardware and not the device being used. I wrote a little [C program](http://pastebin.com/pzx78BwQ) to get the functionality available on a Raspberry Pi.

The mask turns out to be `0xeff0009`, which includes the following -

```
#define I2C_FUNC_I2C            0x00000001                                                                                                                                                                                         
#define I2C_FUNC_SMBUS_PEC      0x00000008                                                                                                                                                                                         
#define I2C_FUNC_SMBUS_QUICK        0x00010000                                                                                                                                                                                     
#define I2C_FUNC_SMBUS_READ_BYTE    0x00020000                                                                                                                                                                                     
#define I2C_FUNC_SMBUS_WRITE_BYTE   0x00040000                                                                                                                                                                                     
#define I2C_FUNC_SMBUS_READ_BYTE_DATA   0x00080000                                                                                                                                                                                 
#define I2C_FUNC_SMBUS_WRITE_BYTE_DATA  0x00100000                                                                                                                                                                                 
#define I2C_FUNC_SMBUS_READ_WORD_DATA   0x00200000                                                                                                                                                                                 
#define I2C_FUNC_SMBUS_WRITE_WORD_DATA  0x00400000                                                                                                                                                                                 
#define I2C_FUNC_SMBUS_PROC_CALL    0x00800000                                                                                                                                                                                                                                                                                                                                                                      
#define I2C_FUNC_SMBUS_WRITE_BLOCK_DATA 0x02000000                                                                                                                                                                                 
#define I2C_FUNC_SMBUS_READ_I2C_BLOCK   0x04000000 /* I2C-like block xfer  */                                                                                                                                                      
#define I2C_FUNC_SMBUS_WRITE_I2C_BLOCK  0x08000000 /* w/ 1-byte reg. addr. */ 
```

Notably absent is -

```
#define I2C_FUNC_SMBUS_READ_BLOCK_DATA  0x01000000
```

On a CHIP, the mask is 0xeff000b (for `/dev/i2c-1`); I2C_FUNC_SMBUS_READ_BLOCK_DATA is absent there as well.

I ran an experiment to read and write using basic read/write syscalls on the file descriptor and it seems to be pretty reliable. Here's the C code to do that on the Raspberry Pi -

```C
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <linux/i2c-dev.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>

#define WII_ADDR 0x52

int main()
{
        int i, fd;
        char init_seq[] = {0x40, 0x00};
        char reset_seq[] = {0x00};
        char buf[6];
        bzero(buf, sizeof(buf));

        if ((fd = open("/dev/i2c-0", O_RDWR)) < 0) {
                fprintf(stderr, "open failed: %s\n", strerror(errno));
                exit(1);
        }
        if (ioctl(fd, I2C_SLAVE, WII_ADDR) < 0) {
                fprintf(stderr, "ioctl failed: %s\n", strerror(errno));
                exit(1);
        }
        if (write(fd, init_seq, sizeof(init_seq)) < 0) {
                fprintf(stderr, "write init sequence failed: %s\n", strerror(errno));
                exit(1);
        }
        usleep(10000);  // let the wiichuck recover
        if (write(fd, reset_seq, sizeof(reset_seq)) < 0) {
                fprintf(stderr, "write reset sequence failed: %s\n", strerror(errno));
                exit(1);
        }
        usleep(10000);  // let the wiichuck recover
        if (read(fd, buf, sizeof(buf)) < sizeof(buf)) {
                fprintf(stderr, "read failed: %s\n", strerror(errno));
                exit(1);
        }

        printf("Wiichuck data: ");
        for (i = 0; i < sizeof(buf); i++) {
                printf("%x ", buf[i]);
        }
        printf("\n");
        return 0;
}
```

Example run -

```
$ gcc -Wall basic_io.c && ./a.out
Wiichuck data: 70 71 81 7f 69 3 
```

So we can fix this by either always doing direct read/write calls to the file descriptor like in the code above, or by querying for functionality first and falling back to the above method if SMBus block read/write is not available. What do you think @deadprogram ?

**Comment by hrishikesh195** [comment]
Resolved by commit 4fdcb6e71850458134aec567c1507a6ba649283e. Closing the issue.
